include/pybind11/common.h

```diff
@@ -30,6 +30,12 @@
 #  define PYBIND11_NOINLINE __attribute__ ((noinline))
 #endif
 
+#if __cplusplus > 201103L
+#  define PYBIND11_DEPRECATED(reason) [[deprecated(reason)]]
+#else
```

Comment:
A useful addition here would be to add:
```C++
#elif defined(__clang__)
#define PYBIND11_DEPRECATED(reason) __attribute__((deprecated(reason)))
#elif defined(__GNUG__)
#define PYBIND11_DEPRECATED(reason) __attribute__((deprecated))
#elif defined(_MSC_VER)
#define PYBIND11_DEPRECATED(reason) __declspec(deprecated)
```
for getting deprecation warning in pre-C++14 compilation as well (though only clang supports also giving a reason).